packages/core/contracts/financial-templates/optimistic-rewarder/OptimisticRewarder.sol

```diff
@@ -0,0 +1,429 @@
+        bytes32 redemptionId = _redemptionId(tokenId, cumulativeRedemptions);
+        require(redemptions[redemptionId].expiryTime == 0, "Redemption already exists");
+        require(ownerOf(tokenId) != address(0), "tokenId is invalid");
+
+        uint256 time = getCurrentTime();
+
+        uint256 finalFee = _getStore().computeFinalFee(address(bondToken)).rawValue;
+        bondToken.safeTransferFrom(msg.sender, address(this), finalFee + bond);
+
+        redemptions[redemptionId] = Redemption({ finalFee: finalFee, expiryTime: time + liveness });
+
+        emit Submitted(tokenId, redemptionId, cumulativeRedemptions, time);
+    }
+
+    /**
+     * @notice Disputes a redemption request.
+     * @dev this will cancel a request if the final fee changes or something causes the optimistic oracle proposal to
+     * fail.
+     * @param tokenId the tokenId the redemption is for.
+     * @param cumulativeRedemptions the cumulative redemptions that were provided in the original request.
+     */
+    function dispute(uint256 tokenId, RedemptionAmount[] memory cumulativeRedemptions) public nonReentrant {
+        bytes32 redemptionId = _redemptionId(tokenId, cumulativeRedemptions);
+
+        // This automatically checks that redemptions[redemptionId] != 0.
+        // Check that it has not passed liveness liveness.
+        Redemption storage redemption = redemptions[redemptionId];
+        require(getCurrentTime() < redemption.expiryTime, "redemption expired or nonexistent");
+        if (redemption.finalFee != _getStore().computeFinalFee(address(bondToken)).rawValue) {
+            _cancelRedemption(tokenId, redemptionId);
+            return;
+        } else {
+            SkinnyOptimisticOracleInterface optimisticOracle = _getOptimisticOracle();
+            uint256 totalBond = bond + redemption.finalFee;
+            bondToken.safeApprove(address(optimisticOracle), totalBond * 2);
+            bytes memory ancillaryData =
+                AncillaryData.appendKeyValueBytes32(customAncillaryData, "redemptionId", redemptionId);
+            uint32 requestTimestamp = uint32(redemption.expiryTime - liveness);
+            address proposer = ownerOf(tokenId);
+
+            try
+                optimisticOracle.requestAndProposePriceFor(
+                    identifier,
+                    requestTimestamp,
+                    ancillaryData,
+                    bondToken,
+                    // Reward = 0
+                    0,
+                    // Set the proposer and disputer bond for the price request.
+                    bond,
+                    // Set the liveness for the price request.
+                    liveness,
+                    proposer,
+                    // Canonical value representing "True"; i.e. the proposed redemption is valid.
+                    int256(1e18)
+                )
+            returns (uint256) {} catch {
```

Comment:
in what case would the catch be hit? if the price request had previously been requested & proposed for for one but are there other cases? we should comment these as this is a bit complex.

Reply:
Sounds good. I will add comments here.

Reply:
Done.